build_tools/benchmarks/xla/benchmark_xla.sh

```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# Benchmarks XLA workloads.
+# Expects environment variables `TENSORFLOW_VERSION` and `RUN_HLO_MODULE_PATH` to be set.
+#
+# Usage:
+#    ./benchmark_xla.sh <tensorflow version> <results dir>
+
+set -xeuo pipefail
+
+export TENSORFLOW_VERSION=$1
+export RESULTS_DIR=$2
+
+# Build Tensorflow from source.
+sudo wget https://github.com/bazelbuild/bazelisk/releases/download/v1.11.0/bazelisk-linux-amd64 -O /usr/local/bin/bazel 
+sudo chmod +x /usr/local/bin/bazel
+
+export TF_BRANCH=$(echo ${TENSORFLOW_VERSION} | sed "s/\([[:digit:]]\+\.[[:digit:]]\+\)\..*/\1/g")
+git clone https://github.com/tensorflow/tensorflow.git
+pushd tensorflow
+git checkout "r${TF_BRANCH}"
+bazel build -c opt --config=cuda tensorflow/compiler/xla/tools/run_hlo_module
```

Comment:
We do plan on breaking the benchmarking infrastructure out into a separate repository, but we don't want to block comparative benchmarking of IREE and XLA:GPU until then. Are there any technical reasons that we should avoid doing this in the meantime, or is it more of a project/repo organization concern?

Reply:
> Are there any technical reasons that we should avoid doing this in the meantime, or is it more of a project/repo organization concern?

Yes, building TensorFlow from source is something we've gone through great pains to distance ourselves from. It's a huge maintenance burden, uses excessive CI resources, and tends to leak into all parts of a project (see the Dockerfile changes here).

Reply:
@mariecwhite, I know you tend to favor avoiding building from source for benchmarking. Any reason that you're opting to build from source here?

FWIW, we're designing these comparative benchmarks to fail gracefully and not take down all of CI -- just not produce benchmarks for that particular target.

Reply:
+1 to isolating XLA:GPU path as much as possible, including homing it in iree-samples or elsewhere.

Reply:
We build TF from source to run compiler-level benchmarks. There is a benchmark binary that will run HLO inputs. For framework-level benchmarks, we use a pip package. We will probably be running this weekly or when a new version of TF is released so we don't expect this to take many CI resources. Maintenance should be minimal if we are using official releases.

Moving this into iree-samples will involve copy/pasting a lot of common code. It also uses `setup.yml`, `./build_tools/github_actions/docker_run.sh` and the helpers that create the docker images. I would rather not duplicate this unless we want to iree-samples to house a prototype for a separate OOBI repo. @pzread How would you like to stage this out?

Reply:
Also, if we want to run these benchmarks on the same machine and time as the IREE benchmarks, we'll need to include it here. I was planning to call this workflow from the `Benchmark Long` workflow.

Reply:
I think the only two things it use from `setup.yml` is runner-group and environment. We can probably just hard-code them like how XLA is doing now https://github.com/openxla/xla/blob/main/.github/workflows/benchmark.yml#L26-L30

Some parts in `docker_run.sh` are also possible to remove (like setting up ccache). It seems like XLA has a simplified command to run docker https://github.com/openxla/xla/blob/main/.github/workflows/benchmark.yml#L43-L47

It makes sense to me to duplicate Dockerfile to build a docker image, because they are built from different purposes. We can also reuse an uploaded IREE nvidia-bleeding-edge image as the base to build a new image for XLA. This introduces a strange dependency on IREE, but probably a quick temporary workaround before we have common base images in OOBI repo

Reply:
I've expressed concern about changes relating to cross-project benchmarks landing in the central IREE repository a few times before, e.g. https://github.com/openxla/iree/pull/11483#pullrequestreview-1214006311. The complexity and maintenance costs have been creeping up, and at a certain point we'll need to pay down that tech debt.

I'd like to see a repository with structure like https://github.com/mlcommons/inference/tree/master/language/bert set up for this work, rather than keep expanding on the Dockerfiles, CI workflows, scripts, etc. in this repo. Whether that repository is an existing ML framework benchmark repository out in the community like those in mlcommons, is some new repo in the openxla GitHub org (`openxla/benchmarks`? `openxla/framework-comparisons`?), or is just `iree-org/iree-samples` is up to you all - but `openxla/iree` is not the place for building and running other compilers/frameworks/runtimes.

Reply:
If we want to benchmark them with IREE, before we implement the proper OOBI repos, we can either:
1. Checkout IREE into iree-sample and run benchmarks there. IREE benchmark suite can be run outside IREE CI, so we should be able to set up that.
2. Run them separately for now and ensure people are aware that they are running on different machines.

To me option 1. is kinda the right direction to move toward the eventual OOBI structure.

Reply:
I'll look into moving this into iree-samples. I've also lodged an issue to include `run_hlo_module` in TF releases: https://github.com/tensorflow/tensorflow/issues/60412

Reply:
> I've expressed concern about changes relating to cross-project benchmarks landing in the central IREE repository a few times before, e.g. [#11483 (review)](https://github.com/openxla/iree/pull/11483#pullrequestreview-1214006311). The complexity and maintenance costs have been creeping up, and at a certain point we'll need to pay down that tech debt.
> 
> I'd like to see a repository with structure like https://github.com/mlcommons/inference/tree/master/language/bert set up for this work, rather than keep expanding on the Dockerfiles, CI workflows, scripts, etc. in this repo. Whether that repository is an existing ML framework benchmark repository out in the community like those in mlcommons, is some new repo in the openxla GitHub org (`openxla/benchmarks`? `openxla/framework-comparisons`?), or is just `iree-org/iree-samples` is up to you all - but `openxla/iree` is not the place for building and running other compilers/frameworks/runtimes.

We are aware of these concerns and have a long-term road map to address this. Let's discuss through the design doc that I shared with you.

Reply:
> I'll look into moving this into iree-samples. I've also lodged an issue to include `run_hlo_module` in TF releases: [tensorflow/tensorflow#60412](https://github.com/tensorflow/tensorflow/issues/60412)

`run_hlo_module` is also available in openxla/xla so this may be easier to build from source.

Reply:
> We are aware of these concerns and have a long-term road map to address this. Let's discuss through the design doc that I shared with you.

Happy to discuss ways to unblock that work and achieve the original goals, but I do want to draw a hard line here - the IREE repository is not the right place for building other projects from source or for running benchmarks for other projects.

Reply:
@mariecwhite can you evaluate what are options are and report back? (doesn't need to be in GH Issues). If there's something we can do to address Scott's concerns that is aligned with our longer-path and is no more than a week of overhead, we can consider it; otherwise, I think the urgency of generating benchmark reliable comparative benchmark results will need to trump ideals of what does/does not belong in the repo. I suspect we'll have a fair amount of temporary consolidation of non-IREE-specific stuff in the IREE repo until all of the CI and benchmarking gets broken out appropriately. (Another example is CI and benchmarking of the openxla-nvgpu compiler.)

Reply:
I'm with Scott on this one. Building another compiler from source does not belong in the IREE repo. A script exclusively benchmarking XLA would make much more sense in the XLA repo. iree-samples doesn't really make much sense either, but the bar for inclusion there is lower. If things currently in the IREE repo are of use for XLA benchmarking then the XLA repository could consider taking a dependency on IREE and would need to evaluate whether that dependency was worth it. Personally, I wouldn't: this is just using a couple of scripts, not even the IREE benchmarking infrastructure. In the meantime, we are actively working on moving the IREE benchmarking infrastructure and other infrastructure into separate repositories (requires an RFC).

From the code reuse perspective, reusable workflows via `workflow_call` can be [used from any repo](https://docs.github.com/en/actions/using-workflows/reusing-workflows#calling-a-reusable-workflow).

Reply:
We will also soon need to benchmark other frameworks in this comparative benchmarking system, including PyTorch and NVIDIA Nemo. We won't be able to escape the reality that we need to invest in integrating other frameworks into our benchmarking infra. We should figure out what the home for those integrations in, and make sure they are modular and fail gracefully without taking down the whole build / system. As I mentioned, let's explore if we can find some expedient ways to avoid doing this for XLA in the IREE repo (even as a temporary solution). But if not, I can't block comparative benchmarking for > ~weeks.

Reply:
The runners are not available in iree-samples. @GMNGeoffrey is it possible to add them? If not, we'll need to run this worklow in this repo.